km/km_filesys.c

```diff
@@ -1951,27 +1962,55 @@ static inline void do_full_write(int fd, char* bufp, size_t bufl)
 static inline size_t
 fs_core_save_pipe_contents(char* buf, size_t length, km_file_t* file, int writefd, size_t queuedbytes)
 {
+   int fd = km_filet2fd(file);
    km_assert(queuedbytes <= length);
+   km_assert(fd >= 0);
 
-   // Read the bytes queued in the pipe
-   do_full_read(file->ofd, buf, queuedbytes);
+   // Read the bytes queued in the pipe into the memory buf points to
+   do_full_read(fd, buf, queuedbytes);
 
-   // now write the bytes back into the pipe
-   do_full_write(writefd, buf, queuedbytes);
+   if (writefd < 0) {
+      // The write side of this pipe is closed.  So we need to recreate the pipe
+      // to put the data back into the pipe buffer and then close the write end
+      // after writing the data.
+      close(fd);
+      int fdpair[2];
+      if (file->how == KM_FILE_HOW_SOCKETPAIR0 || file->how == KM_FILE_HOW_SOCKETPAIR1) {
+         if (socketpair(file->sockinfo->domain, file->sockinfo->type, file->sockinfo->protocol, fdpair) <
+             0) {
+            km_err(1, "Unable to recreate half open socketpair, fd %d", fd);
+         }
+      } else {
+         if (pipe2(fdpair, file->flags) < 0) {
+            km_err(1, "Unable to recreate half open pipe, fd %d", fd);
+         }
+      }
+      do_full_write(fdpair[1], buf, queuedbytes);
+      close(fdpair[1]);
+      if (dup2(fdpair[0], fd) < 0) {
```

Comment:
We need to handle the case where the new pipe/socketpair gets the same fd we started with.  In that case the dup() call is pointless and the close(fdpair[0]) is not needed.